### PR TITLE
fix concurrent accesses to sm_execctrl and sm_instr when sideset isn't optional

### PR DESCRIPTION
**Pull request by ithinuel**
This also includes an alternative implementation of the install function with the hope that using pattern matching will give more opportunity to the compiler to prune the branches we are not interested in.

### CODE REVIEW

File: rp2040-hal/src/pio.rs
Comment (jannic) [inline rp2040-hal/src/pio.rs:543]: I don't like all those added critical sections.
Most of those instances are already guarded from concurrent accesses by the fact that the functions take a mutable reference to the state machine.

The only exception I found is `Tx::drain_fifo(&mut self)`. `Tx` doesn't contain a reference to the state machine, and the write access is realized with help of an unsafe access in `Tx::register_block(&self)`.
What about moving `drain_fifo` to the state machine, or alternatively adding a `&mut StateMachine` parameter, to guarantee exclusive access. That would make the critical sections unnecessary.

BTW, this safety comment is very sus, as the same function appears in both `Tx` and `Rx`:
```
    fn register_block(&self) -> &pac::pio0::RegisterBlock {
        // Safety: The register is unique to this Tx instance.
        unsafe { &*self.block }
    }
```

Comment (ithinuel) [reply rp2040-hal/src/pio.rs:543]: Yes, that's the only reason we need those critical sections. I was tempted to make those calls non-mut but I didn't like that.

It seemed weird to have the drain fifo for tx on the SM without one for Rx, but I'm not against it.
Do you think we should go this way to get rid of the critical sections ?

Comment (jannic) [reply rp2040-hal/src/pio.rs]: Yes, I very much prefer the solution without the critical sections.
While the additional overhead is probably not significant for those functions, critical sections are a global resource: They could affect other users of critical sections, and add noticeable interrupt latency.
I think any critical section in library code should be avoided if possible, and if really needed, they should be limited to code which usually only runs once during initialization, or be as short as possible (eg. a read/modify/write operation).

File: rp2040-hal/src/pio.rs
Comment (jannic) [inline rp2040-hal/src/pio.rs:1228]: Just removing the comment doesn't make the code correct :-)
(But I'll have a closer look if there's really something unsafe here, and if necessary, open a ticket about it. That's not related to this PR)

File: Cargo.toml
Comment (jannic) [inline Cargo.toml:31]: I guess the plan is to remove this as soon as the branch is merged?

Comment (ithinuel) [reply Cargo.toml:31]: Yes

File: rp2040-hal/src/pio.rs
Comment (ithinuel) [reply rp2040-hal/src/pio.rs:494]: Yes, `fn sm(…)`, `fn pio(…)` and `fn register_block(…)` all serve the same purpose. Safety cannot be documented at that point because it depends on what operation is actually done on the registers.

having them all as `unsafe fn` seems fair to me.

**Review by jannic — APPROVED** [review]
Thanks, @ithinuel, for all the work you spent on answering my review comments.
Documenting all the unsafe accesses individually is a huge improvement.

**Pull request merged**
